### ISSUE TITLE
libedgetpu.1.dylib not found

### ISSUE THREAD

**Issue by vladmarian20005** [google-coral/edgetpu#568]
### Description

Hello,

 I've come across an issue while trying to get started with the Coral USB accelerator. When I've tried running the example code it gave me an error saying that "/usr/lib/libedgetpu.1.dylib" does not exist and it truly does not. How can I correctly install it?

<details><summary>Click to expand!</summary> 
 
 ### Issue Type

Build/Install

### Operating System

Mac OS

### Coral Device

USB Accelerator

### Programming Language

Python 3.9

### Relevant Log Output

```shell
Traceback (most recent call last):
  File "/Users/myusername/edgetpu_runtime/coral/pycoral/examples/classify_image.py", line 40, in <module>
    from pycoral.utils.edgetpu import make_interpreter
  File "/Library/Frameworks/Python.framework/Versions/3.9/lib/python3.9/site-packages/pycoral/utils/edgetpu.py", line 24, in <module>
    from pycoral.pybind._pywrap_coral import GetRuntimeVersion as get_runtime_version
ImportError: dlopen(/Library/Frameworks/Python.framework/Versions/3.9/lib/python3.9/site-packages/pycoral/pybind/_pywrap_coral.cpython-39-darwin.so, 0x0002): Library not loaded: @rpath/libedgetpu.1.dylib
  Referenced from: /Library/Frameworks/Python.framework/Versions/3.9/lib/python3.9/site-packages/pycoral/pybind/_pywrap_coral.cpython-39-darwin.so
  Reason: tried: '/Library/Frameworks/Python.framework/Versions/3.9/lib/python3.9/site-packages/pycoral/pybind/libedgetpu.1.dylib' (no such file), '/Library/Frameworks/Python.framework/Versions/3.9/lib/python3.9/site-packages/pycoral/pybind/libedgetpu.1.dylib' (no such file), '/usr/lib/libedgetpu.1.dylib' (no such file)
```
</details>

**Comment by hjonnala** [comment]
Can you please uninstall pycoral and install it using instructions at: https://coral.ai/docs/accelerator/get-started/#pycoral-on-linux. Thanks!

**Comment by vladmarian20005** [comment]
Just uninstalled and reinstalled pycoral for mac and i get the same error! Here is what i have in the pybind folder:
<img width="1304" alt="Screenshot 2022-03-31 at 18 48 51" src="https://user-images.githubusercontent.com/78913181/161096967-7ac584ff-961f-4366-9be4-1677bf5147c4.png">

**Comment by hjonnala** [comment]
Okay, edgetpu runtime seems to be missing. Please install it using instructions at: https://coral.ai/docs/accelerator/get-started/#runtime-on-mac

**Comment by vladmarian20005** [comment]
I had just reinstalled edgetpu and ran the code again. I still receive the same error.

Also, I am able to see some files related to edge tpu so i do not belive that this is the issue

**Comment by hjonnala** [comment]
I have libedgetpu.1.dylib in /usr/local/lib/ directory.

can you please share the output of below commands:
- ls -al  /usr/lib/libedgetpu*
- ls -al /usr/local/lib/libedgetpu*

**Comment by vladmarian20005** [comment]
Definitely!

When I run **ls -al /usr/lib/libedgetpu*** it gives me 'zsh: no matches found: /usr/lib/libedgetpu*'

and when i run **ls -al /usr/local/lib/libedgetpu*** it gives me this :

-rwxr-xr-x  1 root  admin  1024408 Mar 31 18:56 /usr/local/lib/libedgetpu.1.0.dylib
lrwxr-xr-x  1 root  admin       20 Mar 31 18:56 /usr/local/lib/libedgetpu.1.dylib -> libedgetpu.1.0.dylib

**Comment by hjonnala** [comment]
> -rwxr-xr-x 1 root admin 1024408 Mar 31 18:56 /usr/local/lib/libedgetpu.1.0.dylib
 lrwxr-xr-x 1 root admin 20 Mar 31 18:56 /usr/local/lib/libedgetpu.1.dylib -> libedgetpu.1.0.dylib

can you please try the demo copying these files to /usr/lib/.

**Comment by vladmarian20005** [comment]
I have tried moving the files like this:

sudo cp /usr/local/lib/libedgetpu* /usr/lib
Password:
cp: /usr/lib/libedgetpu.1.0.dylib: Operation not permitted
cp: /usr/lib/libedgetpu.1.dylib: Operation not permitted

Seems like I am unable to move or copy them to usr/lib

**Comment by hjonnala** [comment]
Okay, can you try this script by passing any edgetpu model path to the script:: https://github.com/hjonnala/snippets/blob/main/coral_inference.py

`python3 coral_inference.py --model ../../pycoral/test_data/mobilenet_v2_1.0_224_inat_bird_quant_edgetpu.tflite`

changing: 
```
_EDGETPU_SHARED_LIB = {
  'Linux': 'libedgetpu.so.1',
  'Darwin': 'libedgetpu.1.dylib',
  'Windows': 'edgetpu.dll'
}[platform.system()]

```
to

```
_EDGETPU_SHARED_LIB = {
  'Linux': 'libedgetpu.so.1',
  'Darwin': '/usr/local/lib/libedgetpu.1.dylib',
  'Windows': 'edgetpu.dll'
}[platform.system()]

```

**Comment by vladmarian20005** [comment]
It asks me for a model:

usage: tempCodeRunnerFile.python [-h] -m MODEL
tempCodeRunnerFile.python: error: the following arguments are required: -m/--model

**Comment by hjonnala** [comment]
Please pass any edgetpu model path to the script:
` python3 coral_inference.py --model /pycoral/test_data/mobilenet_v2_1.0_224_inat_bird_quant_edgetpu.tflite`

**Comment by vladmarian20005** [comment]
this is the error I receive:

Traceback (most recent call last):
  File "/Users/vladmarian/coral/pycoral/examples/coral_inference.py", line 25, in <module>
    model_path=model_path, experimental_delegates=[tflite.load_delegate(_EDGETPU_SHARED_LIB,  {})])
  File "/Library/Frameworks/Python.framework/Versions/3.9/lib/python3.9/site-packages/tflite_runtime/interpreter.py", line 160, in load_delegate
    delegate = Delegate(library, options)
  File "/Library/Frameworks/Python.framework/Versions/3.9/lib/python3.9/site-packages/tflite_runtime/interpreter.py", line 89, in __init__
    self._library = ctypes.pydll.LoadLibrary(library)
  File "/Library/Frameworks/Python.framework/Versions/3.9/lib/python3.9/ctypes/__init__.py", line 452, in LoadLibrary
    return self._dlltype(name)
  File "/Library/Frameworks/Python.framework/Versions/3.9/lib/python3.9/ctypes/__init__.py", line 374, in __init__
    self._handle = _dlopen(self._name, mode)
OSError: dlopen(libedgetpu.1.dylib, 0x0006): tried: 'libedgetpu.1.dylib' (no such file), '/usr/lib/libedgetpu.1.dylib' (no such file), '/Users/vladmarian/coral/pycoral/libedgetpu.1.dylib' (no such file), '/usr/lib/libedgetpu.1.dylib' (no such file)
Exception ignored in: <function Delegate.__del__ at 0x1079e79d0>
Traceback (most recent call last):
  File "/Library/Frameworks/Python.framework/Versions/3.9/lib/python3.9/site-packages/tflite_runtime/interpreter.py", line 124, in __del__
    if self._library is not None:
AttributeError: 'Delegate' object has no attribute '_library'

**Comment by hjonnala** [comment]
I am not sure why the script is looking for these paths(OSError: dlopen(libedgetpu.1.dylib, 0x0006): tried: 'libedgetpu.1.dylib' (no such file), '/usr/lib/libedgetpu.1.dylib' (no such file), '/Users/vladmarian/coral/pycoral/libedgetpu.1.dylib' (no such file), '/usr/lib/libedgetpu.1.dylib' (no such file)) even though we explicitly define the path. 

Can you please change this line 
> model_path=model_path, experimental_delegates=[tflite.load_delegate(_EDGETPU_SHARED_LIB, {})]) 

with
`model_path=model_path, experimental_delegates=[tflite.load_delegate('/usr/local/lib/libedgetpu.1.dylib', {})]) `. I want to make sure we are providing the path there.

**Comment by vladmarian20005** [comment]
I have just done! This is what it gives me:

6.13 ms

I believe that this is what it is supposed to give me. Is it?

**Comment by hjonnala** [comment]
yes, please go to this file: 
`/Library/Frameworks/Python.framework/Versions/3.9/lib/python3.9/site-packages/pycoral/utils/edgetpu.py` and change these lines (https://github.com/google-coral/pycoral/blob/master/pycoral/utils/edgetpu.py#L34) as below..It would resolve your issue..

```
_EDGETPU_SHARED_LIB = {
  'Linux': 'libedgetpu.so.1',
  'Darwin': '/usr/local/lib/libedgetpu.1.dylib',
  'Windows': 'edgetpu.dll'
}[platform.system()]

```

**Comment by vladmarian20005** [comment]
Just did! Seems like I have a similar issue! It still looks for libedgetpu.1.dylib in usr/lib. I made sure I have saved the modified file.

Traceback (most recent call last):
  File "/Users/vladmarian/coral/pycoral/examples/classify_image.py", line 40, in <module>
    from pycoral.utils.edgetpu import make_interpreter
  File "/Library/Frameworks/Python.framework/Versions/3.9/lib/python3.9/site-packages/pycoral/utils/edgetpu.py", line 24, in <module>
    from pycoral.pybind._pywrap_coral import GetRuntimeVersion as get_runtime_version
ImportError: dlopen(/Library/Frameworks/Python.framework/Versions/3.9/lib/python3.9/site-packages/pycoral/pybind/_pywrap_coral.cpython-39-darwin.so, 0x0002): Library not loaded: @rpath/libedgetpu.1.dylib
  Referenced from: /Library/Frameworks/Python.framework/Versions/3.9/lib/python3.9/site-packages/pycoral/pybind/_pywrap_coral.cpython-39-darwin.so
  Reason: tried: '/Library/Frameworks/Python.framework/Versions/3.9/lib/python3.9/site-packages/pycoral/pybind/libedgetpu.1.dylib' (no such file), '/Library/Frameworks/Python.framework/Versions/3.9/lib/python3.9/site-packages/pycoral/pybind/libedgetpu.1.dylib' (no such file), '/usr/lib/libedgetpu.1.dylib' (no such file)

**Comment by hjonnala** [comment]
Okay, I see in this [script](https://github.com/hjonnala/snippets/blob/main/coral_inference.py) I haven't used pycoral. You might have to work with tflite_runtime for now.

Can you please share the output of below commands:
- uname -a
- sw_vers

**Comment by hjonnala** [comment]
Can you please try to revert those changes and try the command below. Thanks!

- sudo cp -r /usr/local/lib/libedgetpu*  /Library/Frameworks/Python.framework/Versions/3.9/lib/python3.9/site-packages/pycoral/pybind/

**Comment by vladmarian20005** [comment]
> * uname -a
> * sw_vers

uname -a
Darwin 192-168-0-179.rdsnet.ro 21.3.0 Darwin Kernel Version 21.3.0: Wed Jan  5 21:37:58 PST 2022; root:xnu-8019.80.24~20/RELEASE_X86_64 x86_64
vladmarian@192-168-0-179 ~ % sw_vers
ProductName:	macOS
ProductVersion:	12.2.1
BuildVersion:	21D62

**Comment by vladmarian20005** [comment]
> * sudo cp -r /usr/local/lib/libedgetpu*  /Library/Frameworks/Python.framework/Versions/3.9/lib/python3.9/site-packages/pycoral/pybind/

this is the error i get when trying to run the example code:

`Traceback (most recent call last):
  File "/Users/vladmarian/coral/pycoral/examples/classify_image.py", line 121, in <module>
    main()
  File "/Users/vladmarian/coral/pycoral/examples/classify_image.py", line 71, in main
    interpreter = make_interpreter(*args.model.split('@'))
  File "/Library/Frameworks/Python.framework/Versions/3.9/lib/python3.9/site-packages/pycoral/utils/edgetpu.py", line 87, in make_interpreter
    delegates = [load_edgetpu_delegate({'device': device} if device else {})]
  File "/Library/Frameworks/Python.framework/Versions/3.9/lib/python3.9/site-packages/pycoral/utils/edgetpu.py", line 52, in load_edgetpu_delegate
    return tflite.load_delegate(_EDGETPU_SHARED_LIB, options or {})
  File "/Library/Frameworks/Python.framework/Versions/3.9/lib/python3.9/site-packages/tflite_runtime/interpreter.py", line 160, in load_delegate
    delegate = Delegate(library, options)
  File "/Library/Frameworks/Python.framework/Versions/3.9/lib/python3.9/site-packages/tflite_runtime/interpreter.py", line 89, in __init__
    self._library = ctypes.pydll.LoadLibrary(library)
  File "/Library/Frameworks/Python.framework/Versions/3.9/lib/python3.9/ctypes/__init__.py", line 452, in LoadLibrary
    return self._dlltype(name)
  File "/Library/Frameworks/Python.framework/Versions/3.9/lib/python3.9/ctypes/__init__.py", line 374, in __init__
    self._handle = _dlopen(self._name, mode)
OSError: dlopen(libedgetpu.1.dylib, 0x0006): tried: 'libedgetpu.1.dylib' (no such file), '/usr/lib/libedgetpu.1.dylib' (no such file), '/Users/vladmarian/coral/pycoral/libedgetpu.1.dylib' (no such file), '/usr/lib/libedgetpu.1.dylib' (no such file)
Exception ignored in: <function Delegate.__del__ at 0x111af8b80>
Traceback (most recent call last):
  File "/Library/Frameworks/Python.framework/Versions/3.9/lib/python3.9/site-packages/tflite_runtime/interpreter.py", line 124, in __del__
    if self._library is not None:
AttributeError: 'Delegate' object has no attribute '_library'`

**Comment by hjonnala** [comment]
can you please try to copy to pycoral folder as well.

sudo cp -r /usr/local/lib/libedgetpu* /Users/vladmarian/coral/pycoral/

**Comment by vladmarian20005** [comment]
> sudo cp -r /usr/local/lib/libedgetpu* /Users/vladmarian/coral/pycoral/

Wow i have just did and it worked!!! Finally!!

`----INFERENCE TIME----
Note: The first inference on Edge TPU is slow because it includes loading the model into Edge TPU memory.
19.2ms
6.0ms
5.5ms
5.2ms
5.0ms
-------RESULTS--------
Ara macao (Scarlet Macaw): 0.75781`

Thank you so much for the help! I believe that I have nothing left to do! Thanks!!

**Comment by EagleDTW** [comment]
Hi @hjonnala, Trying to get my coral up and running on a MacOS Tahoe and running into extremely similar issues as this use case...could you help me please to diagnose?




Traceback (most recent call last):
  File "examples/classify_image.py", line 121, in <module>
    main()
  File "examples/classify_image.py", line 71, in main
    interpreter = make_interpreter(*args.model.split('@'))
  File "/Users/daniel/.pyenv/versions/3.8.20/lib/python3.8/site-packages/pycoral/utils/edgetpu.py", line 87, in make_interpreter
    delegates = [load_edgetpu_delegate({'device': device} if device else {})]
  File "/Users/daniel/.pyenv/versions/3.8.20/lib/python3.8/site-packages/pycoral/utils/edgetpu.py", line 52, in load_edgetpu_delegate
    return tflite.load_delegate(_EDGETPU_SHARED_LIB, options or {})
  File "/Users/daniel/.pyenv/versions/3.8.20/lib/python3.8/site-packages/tflite_runtime/interpreter.py", line 160, in load_delegate
    delegate = Delegate(library, options)
  File "/Users/daniel/.pyenv/versions/3.8.20/lib/python3.8/site-packages/tflite_runtime/interpreter.py", line 89, in __init__
    self._library = ctypes.pydll.LoadLibrary(library)
  File "/Users/daniel/.pyenv/versions/3.8.20/lib/python3.8/ctypes/__init__.py", line 451, in LoadLibrary
    return self._dlltype(name)
  File "/Users/daniel/.pyenv/versions/3.8.20/lib/python3.8/ctypes/__init__.py", line 373, in __init__
    self._handle = _dlopen(self._name, mode)
OSError: dlopen(libedgetpu.1.dylib, 0x0006): tried: 'libedgetpu.1.dylib' (no such file), '/System/Volumes/Preboot/Cryptexes/OSlibedgetpu.1.dylib' (no such file), '/Users/daniel/.pyenv/versions/3.8.20/lib/libedgetpu.1.dylib' (no such file), '/System/Volumes/Preboot/Cryptexes/OS/Users/daniel/.pyenv/versions/3.8.20/lib/libedgetpu.1.dylib' (no such file), '/opt/homebrew/lib/libedgetpu.1.dylib' (no such file), '/System/Volumes/Preboot/Cryptexes/OS/opt/homebrew/lib/libedgetpu.1.dylib' (no such file), '/Users/daniel/.pyenv/versions/3.8.20/lib/libedgetpu.1.dylib' (no such file), '/System/Volumes/Preboot/Cryptexes/OS/Users/daniel/.pyenv/versions/3.8.20/lib/libedgetpu.1.dylib' (no such file), '/opt/homebrew/lib/libedgetpu.1.dylib' (no such file), '/System/Volumes/Preboot/Cryptexes/OS/opt/homebrew/lib/libedgetpu.1.dylib' (no such file), '/usr/lib/libedgetpu.1.dylib' (no such file, not in dyld cache), 'libedgetpu.1.dylib' (no such file)
Exception ignored in: <function Delegate.__del__ at 0x1072d8b80>
Traceback (most recent call last):
  File "/Users/daniel/.pyenv/versions/3.8.20/lib/python3.8/site-packages/tflite_runtime/interpreter.py", line 124, in __del__
    if self._library is not None:
AttributeError: 'Delegate' object has no attribute '_library'


TIA...all the help appreciated!

**Comment by EagleDTW** [comment]
I used python 3.8 because the modules wouldn't load in 3.9 and were apparently missing from pycoral install in 3.9 but present in 3.8. I certainly can adjust this to what is needed. pyenv installed using homebrew used the project homepage directions to install most everything else get the errors above when trying to run the demo. Just really need the vendor id to change so that I can pass the usb into a UTM VM of HAOS to use for inferencing in there.

I used your above script adjusting the 2 different lines referencing the /usr/local/lib file location and no luck there still have error messages as well, abeit different ones than above - I've included below:

Traceback (most recent call last):
  File "/Users/daniel/.pyenv/versions/3.8.20/lib/python3.8/site-packages/tflite_runtime/interpreter.py", line 160, in load_delegate
    delegate = Delegate(library, options)
  File "/Users/daniel/.pyenv/versions/3.8.20/lib/python3.8/site-packages/tflite_runtime/interpreter.py", line 119, in __init__
    raise ValueError(capture.message)
ValueError

During handling of the above exception, another exception occurred:

Traceback (most recent call last):
  File "coral_inference.py", line 25, in <module>
    model_path=model_path, experimental_delegates=[tflite.load_delegate('/usr/local/lib/libedgetpu.1.dylib', {})])
  File "/Users/daniel/.pyenv/versions/3.8.20/lib/python3.8/site-packages/tflite_runtime/interpreter.py", line 162, in load_delegate
    raise ValueError('Failed to load delegate from {}\n{}'.format(
ValueError: Failed to load delegate from /usr/local/lib/libedgetpu.1.dylib
